lib/kubecfg.libsonnet

```diff
@@ -56,4 +56,55 @@
   // to refer to submatches.  Regex is as implemented in golang regexp
   // package (python-ish).
   regexSubst:: std.native("regexSubst"),
+
+  // isK8sObject(o): Return true iff o is a Kubernetes object.
+  isK8sObject(o):: (
+    std.isObject(o) &&
+    std.objectHas(o, "apiVersion") &&
+    std.objectHas(o, "kind")
+  ),
+
+  // Private helper function for map/fold.
+  local isK8sList(o) = $.isK8sObject(o) && o.apiVersion == "v1" && o.kind == "List",
+
+  // deepMap(func, o): Apply the given function to each Kubernetes
+  // object in nested collection o, preserving the structure of o.
+  deepMap(func, o):: (
+    if isK8sList(o) then
+      o + {items: [func(item) for item in super.items]}
+    else if $.isK8sObject(o) then
+      func(o)
+    else if std.isObject(o) then
+      {[k]: $.deepMap(func, o[k]) for k in std.objectFields(o)}
+    else if std.isArray(o) then
+      [$.deepMap(func, elem) for elem in o]
+    else if o == null then
+      null
+    else
+      error ("o must be an object or array of k8s objects, found " + std.type(o))
+  ),
+
+  // fold(func, o, init): Apply the given function to each Kubernetes
+  // object in nested collection o, accumulating a result as we go.
+  // Function arg is invoked with arguments (accumulator, object).
+  fold(func, o, init):: (
+    if isK8sList(o) then
+      $.fold(func, o.items, init)
+    else if $.isK8sObject(o) then
+      func(init, o)
+    else if std.isObject(o) then
+      $.fold(func, [o[k] for k in std.objectFields(o)], init)
+    else if std.isArray(o) then
+      std.foldl(function (running, elem) $.fold(func, elem, running), o, init)
+    else if o == null then
+      init
+    else
+      error ("o must be an object or array of k8s objects, found " + std.type(o))
+  ),
+
+  // byGvkName(accum, o): Helper for 'fold'.  This accumulates a
+  // two-level collection of objects by 'apiVersion.kind'
+  // (GroupVersionKind) and then object 'name'.
+  byGvkName(accum, o)::
+    accum + {[o.apiVersion + "." + o.kind]+: {[o.metadata.name]: o}},
```

Comment:
Another thing: the name field is not mandatory. Objects could have a `genetatedName` field instead.

Perhaps we should handle that too

Reply:
I agree, which is why this function only works 95% of the time.  What should we do for the generatedName case, since the object doesn't have a universal identifier in jsonnet?

For example, the following is a valid, useful kubernetes manifest.  It has zero way to separate the three jobs, other than by something contextual like position in the resulting array:

```jsonnet
local job = {apiVersion: "batch/v1", kind: "Job", metadata: {generateName: "myjob-"}};
[job, job, job]
```

Fwiw, here's the array index 'name' solution, that is the right thing for this specific case:

```jsonnet
local kubecfg = import "kubecfg.jsonnet";
local byPosition(accum, o) = accum + {["index" + std.toString(std.length(accum))]: o};

local objs = kubecfg.fold(byPosition, jobs, {});
objs + {
  index2+: {
     metadata+: {annotations+: {"foo": "I am an override for the 3rd job"}},
  },
}
```

Generates:
```yaml
---
apiVersion: batch/v1
kind: Job
metadata:
  generateName: myjob-
---
apiVersion: batch/v1
kind: Job
metadata:
  generateName: myjob-
---
apiVersion: batch/v1
kind: Job
metadata:
  annotations:
    foo: I am an override for the 3rd job
  generateName: myjob-
```

I think `fold` **is** the universal powerful function that you're looking for.  The accumulator function we pass to `fold` is by necessity a special case - that's why it's a _parameter_ that the user is meant to provide.

Reply:
yeah; perhaps I'm biased; I rarely see multiple generatedNames with the same prefix in the same resource collection. I see it more used in order to generate new names every time they get applied.

I guess since we cannot cover 100% it's probably better to keep it simple.

Reply:
I agree i never saw multiple objects with the same `generate-name` , the only times i saw the generate name is for jobs to trigger a recreate but is most time just one job

so i'd say we might be covering 99% of use cases already which is great